File: CodingSamples/Java/Language-Runtime/AbstractTypeTest/bank2/src/banking/SavingsAccount.java
package banking;

final class SavingsAccount extends Account implements Profitable {
    
    //value of a field declared with 'static' modifier is shared by
    //all the instances of its class, a field declared with 'final' 
    //modifier cannot be reassinged after it is initialized
    static final double MIN_BAL = 5000;

    SavingsAccount() {
        balance = MIN_BAL;
    }

    public void deposit(double amount) {
        balance += amount;
    }

    public void withdraw(double amount) throws InsufficientFundsException {
        if(balance - amount < MIN_BAL)
            throw new InsufficientFundsException();
        balance -= amount;
    }

    public double interest(int months) {
        double rate = balance < 5 * MIN_BAL ? 3 : 4;
        return balance * months * rate / 1200;
    }

}

File: CodingSamples/DotNet/Language-Runtime/AbstractTypeTest/BankLib2/SavingsAccount.cs
namespace Banking;

sealed class SavingsAccount : Account, IProfitable
{
    const double MinBal = 5000;

    internal SavingsAccount()
    {
        Balance = MinBal;
    }

    public override void Deposit(double amount)
    {
        Balance += amount;
    }

    public override void Withdraw(double amount)
    {
        if(Balance - amount < MinBal)
            throw new InsufficientFundsException();
        Balance -= amount;
    }

    public double AddInterest(int months)
    {
        double rate = Balance < 5 * MinBal ? 3 : 4;
        double interest = Balance * rate * months / 1200;
        Balance += interest;
        return interest;
    }

}